### ISSUE TITLE
Callback API for Trainer

### ISSUE THREAD

**Issue by tomekkorbak** [facebookresearch/EGG#7]
I really love the library, find it extremely well-designed and started using it for some of my own research. One thing that bothers me is lack of extensibility of trainer with respect to monitoring  and learning rate scheduling. I'm thinking, for instance, custom monitors and loggers other than Tensorboard. The `_message` function in `core.Trainer.train` is not accessible outside of the `train` method and the `epoch_callback` is of little use for monitoring because the losses and `rest` dicts accessible via `self`.

Have you thought about implementing a simple callback logic, similar to the callback API found in Keras, fastai or ignite? Would a pull request adding such a thing be welcome? Or do you have some other direction for developing EGG in mind?

**Comment by eugene-kharitonov** [comment]
Hello @tomekkorbak! Learning rate scheduling and other loggers is not something on our short-term list, but indeed this could be useful. 
A PR will be absolutely welcome, however, it would be nice if we could discuss the API changes before an actual PR.

**Comment by tomekkorbak** [comment]
Thanks for your reply, @eugene-kharitonov. I thought about something along these lines
```python
class Callback:
    trainer: Trainer
        
    def on_train_begin(self, trainer_instance: Trainer):
        self.trainer = trainer  # So that each callback can manipulate trainer state, e.g. set trainer.should_stop()
    
    def on_train_end(self):
        pass
        
    def on_test_begin(self):
        pass
    
    def on_test_end(self, loss: float, logs: Dict[str, Any] = None):
        pass
    
    def on_epoch_begin(self, logs: Dict[str, Any] = None):
        pass

    def on_epoch_end(self):
        pass
```

It more or less mirrors callback APIs in fastai and Keras, enabling plugging in reusable code snippets for monitoring and logging.

The `core.Trainer.train` method would have to be rewritten to call these methods for each callback:
```python
    def train(self, n_epochs):
        for callback in self.callbacks:
            callback.on_epoch_begin(self)
        for epoch in range(n_epochs):
            for callback in self.callbacks:
                callback.on_epoch_begin()
            train_loss, train_rest = self.train_epoch()    
            for callback in self.callbacks:
                callback.on_epoch_end(train_loss, train_rest)    

            if self.validation_data is not None and self.validation_freq > 0 and epoch % self.validation_freq == 0:
                for callback in self.callbacks:
                    callback.on_test_begin()
                validation_loss, rest = self.eval()
                for callback in self.callbacks:
                    callback.on_test_end(validation_loss, rest)

                if self.should_stop()
                    break
        for callback in self.callbacks:
            callback.on_train_end()
```

I would move some of the logic currently found in `train` (early stopping, Tensorboard logging, printing and snapshoting) to appropriate callbacks and included some of them by default

```python
def __init__(self, game, optimizer, train_data, validation_data=None, device=None, callbacks=[]):
    # ...
    if callbacks is None:
        self.callbacks = [
            ConsoleLogger(as_json=False),
        ]
    else:
        self.callbacks = callbacks
```

And let create classes for the rest of them to be initialized and passed as arguments during trainer instantiation:
```python
trainer = Trainer(
    *some_args, 
    callbacks=[
        TensorboardLogger(summary_writer=get_summary_writer()),
        EarlyStopperAccuracy(threshold=0.9),
        CheckpointCallback(checkpoint_path=common_opts.checkpoint_dir)
    ]
)
```

It would be easy now to write a custom callback by inheriting from the `Callback` and passing the instance as an argument during trainer instantiation.

Is my sketch of the API clear enough to illustrate how it would work? Note that the above code was not tested and is only mean to illustrate the API rather than be the final implementation. Is this a direction you feel EGG development could pursue?

**Comment by eugene-kharitonov** [comment]
Awesome, thank you so much for a such a detailed proposal! Please give me a day or two to find time for a proper reply.

**Comment by eugene-kharitonov** [comment]
@tomekkorbak sorry for a late reply. 
Indeed, sounds clean and nice. Do you want to implement that yourself or should I do that?

**Comment by tomekkorbak** [comment]
@eugene-kharitonov, I'm glad you liked the API. Sure, I'd be happy to prepare a pull request (at least with the API itself, see below). 

One thing that needs to be discussed is how to partition these changes into pull requests. Implementing the API (i.e. abstract class `Callback` and changes in the `Trainer` class) is one thing and rewriting some of the existing features (early stopping, snapshotting, Tensorboard monitoring, console logging) is yet another. Would you these changes to be split into several pull requests? Would you like the first pull request to keep the existing implementation of (non-callback-based) early stopping etc until another pull request will reimplement it?

**Comment by eugene-kharitonov** [comment]
Hey, I think EGG is small enough so it's should be huge change and both ways are fine to mine. Just bear in mind that some implemented games use early stopping :)

**Comment by tomekkorbak** [comment]
Great, I think I could work on that next week!

**Comment by tomekkorbak** [comment]
I've opened a pull request and I'm looking forward to hear your feedback, @eugene-kharitonov!

**Comment by tomekkorbak** [comment]
I've just opened #17, which completes implementing the functionality mentioned above. This is why I'm closing this issue.
